### ISSUE TITLE
Automatically generate resolve functions

### ISSUE THREAD

**Issue by jsindos** [graphql-python/graphene-django#256]
At the bottom of the official tutorial it says

>As you can see, GraphQL is very powerful but there are a lot of repetitions in our example. We can do a lot of improvements by adding layers of abstraction on top of `graphene-django`.

However, in the 1.0 blog post it says

>Ok, it’s hard to do a maintainable Python package based in the in the syntactic sugar magic behind the metaclasses.

I'm new to Python metaclasses and thought it would be a cool challenge to try and automatically add resolve functions for certain fields. I basically want to make `resolve_*` functions be automatically added in simple cases:

```
class Query(graphene.AbstractType):
    lesson = graphene.Field(LessonType, id=graphene.Int())

    # def resolve_lesson(self, args, context, info):
    #     This will get automatically added!
    #     id = args.get('id')

    #     if slug is not None:
    #         return Lesson.objects.get(id=id)

    #     return None
```

I tried to do this by defining my own metaclass inheriting from `AbstractTypeMeta`, and adding extra resolve functions to attrs before `AbstractTypeMeta` does its thing. Unfortunately due to the way the schemas are combined with `graphene.ObjectType`, I get a metaclass conflict. At this point I gave up as it seems like I'm either missing something or it will be a long headache if I continue.

My question is, are there any capabilities or plans to be able to do what I want to do?

By the way, very excited to start using GraphQL! I've only just discovered it today but it looks very promising.

**Comment by yatriks** [comment]
I've managed to do something like this by creating a metaclass that subclasses `DjangoObjectTypeMeta.__new__(cls, name, parents, dct)` and manipulates `dct` based on properties of `dct['Meta'].model`. 

e.g. 

```
class MyObjectType(DjangoObjectTypeMeta):
    def __new__(cls, name, parents, dct):
        if hasattr(dct['Meta'].model, 'get_url'):
            dct['url'] = graphene.String(resolver=lambda self, *args, **kwargs: self.get_url())
        return super(MyObjectType, cls).__new__(cls, name, parents, dct)

class MyNode(DjangoObjectType):
    __metaclass__ = MyObjectType
    
    class Meta:
        model = MyModel
```

**Comment by spockNinja** [comment]
We are using a sub class of `graphene.Field` to handle this sort of thing.

```python
class CustomModelField(Field):
    """
    A subclass of graphene's builtin Field.
    We automatically find non-id unique attributes and allow those
    as arguments to the given node type.

    Usage:
        entity = CustomModelField(EntityGrapheneType)
    """

    def __init__(self, field_type, **kwargs):
        self.field_type = get_type(field_type)

        unique_fields = self.get_unique_fields(_type=field_type)

        is_required = len(unique_fields) == 1

        graphql_args = {}
        for model_field in unique_fields:
            form_field = model_field.formfield() or CharField()
            form_field.required = is_required
            graphene_field = convert_form_field(form_field)
            graphql_args[model_field.name] = graphene_field

        all_kwargs = graphql_args
        all_kwargs.update(kwargs)

        super().__init__(
            field_type,
            **all_kwargs
        )

    def get_unique_fields(self, _type=None):
        if _type:
            model = self.get_model_from_type(_type)
        else:
            model = self.model

        unique_fields = [f for f in model._meta.local_fields if f.unique]
        return unique_fields

    def model_resolver(self, root, args, context, info):
        if not args:
            unique_field_names = [f.name for f in self.get_unique_fields()]
            err_msg = 'At least one of {} is required.'.format(unique_field_names)
            raise GraphQLError(err_msg)

        qs = optimize_queryset(self.manager, info)

        return self.model._default_manager.get(**args)

    def get_resolver(self, parent_resolver):
        return self.model_resolver
```

I cut out a bit for brevity to just show the idea. Basically, we automatically add arguments based on any unique fields on the model. So from your example, it'd just be:

```python
class Query(graphene.AbstractType):
    lesson = CustomModelField(LessonType)
```

Then the `id` would be added as an argument, and the resolver would just use the ORM to pass the unique fields to `default_manger.get`.

**Comment by shifqu** [comment]
Hello.

Thanks for the input on this ticket. I ended up using a combination of @yatriks and @spockNinja .
The first I implemented to add another field to the objecttype meta and the second I use to create a default custom resolver for the graphene.Field type.

Now I am at a loss by trying to do the same thing as I did for graphene.Field, but this time for a graphene.List. The latter does not seem to accept resolver of get_resolver by default.

Anyone has an idea on how to achieve getting a default_resolver for the graphene.List type?
What I have tried so far:
```python
class BaseList(graphene.List):

    def __init__(self, of_type, *args, **kwargs):
        kwargs.update(of_type._meta.field_matrix)
        super(BaseList, self).__init__(of_type, *args, **kwargs)

    def model_resolver(self, *args, **kwargs):
        ... logic to return a queryset ...

    def get_resolver(self, parent_resolver):
        return self.model_resolver
```
But as I explained before, this does not work for List, it does work for Field
